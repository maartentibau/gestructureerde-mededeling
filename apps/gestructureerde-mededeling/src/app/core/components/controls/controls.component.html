<ng-container *ngIf="isSmallScreen$ | async; then smallControls; else largeControls"></ng-container>

<ng-template #smallControls>
  <div class="controls">
    <button
      *ngIf="refresh"
      mat-fab
      disableRipple
      color="primary"
      type="button"
      title="Verniewen"
      (click)="refreshClickHandler()"
    >
      <fa-icon [icon]="['fas', 'sync-alt']"></fa-icon>
    </button>
    <button
      *ngIf="copyNumber"
      class="ogm-copy"
      mat-fab
      disableRipple
      color="primary"
      type="button"
      title="Kopieer nummer"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm?.number"
      (click)="copyNumberClickHandler()"
    >
      <fa-icon [icon]="['fas', 'comment-dots']"></fa-icon>
    </button>
    <button
      *ngIf="copyOgm"
      class="ogm-copy"
      mat-fab
      disableRipple
      color="primary"
      type="button"
      title="Kopieer mededeling"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm?.numberFormat"
      (click)="copyOgmClickHandler()"
    >
      <fa-icon [icon]="['fas', 'money-check']"></fa-icon>
    </button>
  </div>
</ng-template>

<ng-template #largeControls>
  <div class="controls large">
    <button
      *ngIf="refresh"
      mat-raised-button
      isRoundButton
      disableRipple
      color="primary"
      (click)="refreshClickHandler()"
    >
      <fa-icon [icon]="['fas', 'sync-alt']"></fa-icon>Vernieuwen
    </button>
    <button
      *ngIf="copyNumber"
      class="ogm-copy"
      mat-raised-button
      isRoundButton
      disableRipple
      color="primary"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm?.number"
      (click)="copyNumberClickHandler()"
    >
      <fa-icon [icon]="['fas', 'comment-dots']"></fa-icon>Kopieer nummer
    </button>
    <button
      *ngIf="copyOgm"
      class="ogm-copy"
      mat-raised-button
      isRoundButton
      disableRipple
      color="primary"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm?.numberFormat"
      (click)="copyOgmClickHandler()"
    >
      <fa-icon [icon]="['fas', 'money-check']"></fa-icon>Kopieer mededeling
    </button>
  </div>
</ng-template>
