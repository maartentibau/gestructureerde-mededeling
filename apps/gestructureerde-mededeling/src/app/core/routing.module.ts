import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreateComponent } from '../components/create/create.component';
import { GenerateComponent } from '../components/generate/generate.component';
import { ValidateComponent } from '../components/validate/validate.component';

const routes: Routes = [
  { path: 'create', component: CreateComponent },
  { path: 'generate', component: GenerateComponent },
  { path: 'validate', component: ValidateComponent },
  { path: '', redirectTo: '/generate', pathMatch: 'full' },
  { path: '**', component: GenerateComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { relativeLinkResolution: 'legacy' })],
  exports: [RouterModule],
})
export class RoutingModule {}
