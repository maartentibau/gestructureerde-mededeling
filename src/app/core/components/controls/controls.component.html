@if (isSmallScreen$ | async) {
  <div class="controls">
    @if (refresh()) {
      <button mat-fab disableRipple color="primary" type="button" title="Verniewen" (click)="refreshClickHandler()">
        <fa-icon [icon]="['fas', 'sync-alt']"></fa-icon>
      </button>
    }
    @if (copyNumber()) {
      <button
        class="ogm-copy"
        mat-fab
        disableRipple
        color="primary"
        type="button"
        title="Kopieer nummer"
        data-clipboard-action="copy"
        [attr.data-clipboard-text]="ogm()?.number"
        (click)="copyNumberClickHandler()"
      >
        <fa-icon [icon]="['fas', 'comment-dots']"></fa-icon>
      </button>
    }
    @if (copyOgm()) {
      <button
        class="ogm-copy"
        mat-fab
        disableRipple
        color="primary"
        type="button"
        title="Kopieer mededeling"
        data-clipboard-action="copy"
        [attr.data-clipboard-text]="ogm()?.numberFormat"
        (click)="copyOgmClickHandler()"
      >
        <fa-icon [icon]="['fas', 'money-check']"></fa-icon>
      </button>
    }
  </div>
} @else {
  <div class="controls large">
    @if (refresh()) {
      <button mat-raised-button disableRipple color="primary" (click)="refreshClickHandler()">
        <fa-icon [icon]="['fas', 'sync-alt']"></fa-icon>Vernieuwen
      </button>
    }
    @if (copyNumber()) {
      <button
        class="ogm-copy"
        mat-raised-button
        disableRipple
        color="primary"
        data-clipboard-action="copy"
        [attr.data-clipboard-text]="ogm()?.number"
        (click)="copyNumberClickHandler()"
      >
        <fa-icon [icon]="['fas', 'comment-dots']"></fa-icon>Kopieer nummer
      </button>
    }
    @if (copyOgm()) {
      <button
        class="ogm-copy"
        mat-raised-button
        disableRipple
        color="primary"
        data-clipboard-action="copy"
        [attr.data-clipboard-text]="ogm()?.numberFormat"
        (click)="copyOgmClickHandler()"
      >
        <fa-icon [icon]="['fas', 'money-check']"></fa-icon>Kopieer mededeling
      </button>
    }
  </div>
}
