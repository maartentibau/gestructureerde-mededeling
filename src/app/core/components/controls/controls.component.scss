@import '../../../../assets/styles/screen-media-queries';

.controls {
  margin: 2rem 0.5rem;
  display: flex;
  justify-content: space-around;

  button {
    font-size: 1.5rem;
    background-color: #18b351;

    &:active {
      background-color: #fff;
      color: #18b351;
      cursor: pointer;
    }
  }

  &.large {
    button {
      height: 60px;
      flex: 1 1 0;
      line-height: 60px;
      max-width: 450px;
      margin: 0 16px;

      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;

      fa-icon {
        line-height: 1.5rem;
        padding-right: 12px;
      }

      font-size: 1.5rem;

      @media #{$mat-md} {
        max-width: 250px;
        height: 50px;
        font-size: 18px;
        line-height: 50px;

        mat-icon {
          padding-right: 0.5rem;
        }
      }

      @media #{$mat-lt-md} {
        margin: 0 0.5rem;
        height: 40px;
        font-size: 1rem;
        line-height: 40px;

        mat-icon {
          padding-right: 6px;
        }
      }
    }
  }
}
