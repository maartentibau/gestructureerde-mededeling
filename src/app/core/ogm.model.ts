export interface OgmData {
  number: string;
  numberFormat: string;
}
