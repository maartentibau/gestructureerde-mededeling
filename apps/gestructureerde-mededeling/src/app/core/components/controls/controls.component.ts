import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnDestroy, Output } from '@angular/core';
import { ScreenService } from '../../services/screen.service';
import { FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { map } from 'rxjs/operators';
import { Breakpoints } from '@angular/cdk/layout';
import { Observable } from 'rxjs';
import { faCommentDots, faCommentLines, faSyncAlt } from '@fortawesome/pro-duotone-svg-icons';
import { OgmData } from '../../ogm.model';
import ClipboardJS from 'clipboard';

@Component({
  selector: 'ogm-controls',
  templateUrl: './controls.component.html',
  styleUrls: ['./controls.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ControlsComponent implements OnDestroy {
  @Input() refresh: boolean | undefined;
  @Input() copyNumber: boolean | undefined;
  @Input() copyOgm: boolean | undefined;
  @Input() ogm: OgmData | null | undefined;

  @Output() refreshClick: EventEmitter<void> = new EventEmitter<void>();
  @Output() copyNumberClick: EventEmitter<string> = new EventEmitter<string>();
  @Output() copyOgmClick: EventEmitter<string> = new EventEmitter<string>();

  readonly isSmallScreen$: Observable<boolean>;
  readonly clipboard: ClipboardJS;

  constructor(private screenService: ScreenService, private faIconLibrary: FaIconLibrary) {
    this.isSmallScreen$ = this.screenService
      .observerBreakpoints()
      .pipe(map((breakpoints: { [key: string]: boolean }) => breakpoints[Breakpoints.XSmall]));

    this.clipboard = new ClipboardJS('.ogm-copy');

    this.faIconLibrary.addIcons(faSyncAlt, faCommentDots, faCommentLines);
  }

  ngOnDestroy() {
    this.clipboard.destroy();
  }

  refreshClickHandler() {
    this.refreshClick.emit();
  }

  copyNumberClickHandler() {
    this.copyNumberClick.emit('Nummer gekopieerd naar klembord');
  }

  copyOgmClickHandler() {
    this.copyOgmClick.emit('Mededeling gekopieerd naar klembord');
  }
}
