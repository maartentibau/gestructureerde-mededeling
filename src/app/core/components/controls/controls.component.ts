import { Breakpoints } from '@angular/cdk/layout';
import { AsyncPipe, NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, EventEmitter, inject, Input, OnDestroy, Output } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { FaIconLibrary, FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { faCommentDots, faMoneyCheck, faSyncAlt } from '@fortawesome/free-solid-svg-icons';
import ClipboardJS from 'clipboard';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { OgmData } from '../../ogm.model';
import { ScreenService } from '../../services/screen.service';

@Component({
  standalone: true,
  imports: [NgIf, AsyncPipe, MatButtonModule, FontAwesomeModule],
  selector: 'ogm-controls',
  templateUrl: './controls.component.html',
  styleUrls: ['./controls.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ControlsComponent implements OnDestroy {
  #screenService: ScreenService = inject(ScreenService);
  #faIconLibrary: FaIconLibrary = inject(FaIconLibrary);

  @Input() refresh: boolean | undefined;
  @Input() copyNumber: boolean | undefined;
  @Input() copyOgm: boolean | undefined;
  @Input() ogm: OgmData | null | undefined;

  @Output() refreshClick: EventEmitter<void> = new EventEmitter<void>();
  @Output() copyNumberClick: EventEmitter<string> = new EventEmitter<string>();
  @Output() copyOgmClick: EventEmitter<string> = new EventEmitter<string>();

  readonly isSmallScreen$: Observable<boolean>;
  readonly clipboard: ClipboardJS;

  constructor() {
    this.isSmallScreen$ = this.#screenService
      .observerBreakpoints()
      .pipe(map((breakpoints: { [key: string]: boolean }) => breakpoints[Breakpoints.XSmall]));

    this.clipboard = new ClipboardJS('.ogm-copy');

    this.#faIconLibrary.addIcons(faSyncAlt, faCommentDots, faMoneyCheck);
  }

  ngOnDestroy() {
    this.clipboard.destroy();
  }

  refreshClickHandler() {
    this.refreshClick.emit();
  }

  copyNumberClickHandler() {
    this.copyNumberClick.emit('Nummer gekopieerd naar klembord');
  }

  copyOgmClickHandler() {
    this.copyOgmClick.emit('Mededeling gekopieerd naar klembord');
  }
}
