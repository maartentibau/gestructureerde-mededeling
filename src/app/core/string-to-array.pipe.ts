import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
