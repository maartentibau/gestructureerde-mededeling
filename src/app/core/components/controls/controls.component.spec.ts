import { AsyncPipe, JsonPipe, NgIf } from '@angular/common';
import { Component, Input } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatButtonModule } from '@angular/material/button';
import { FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { ScreenService } from '../../services/screen.service';

import { ControlsComponent } from './controls.component';

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'fa-icon',
  template: ` <div>{{ icon | json }}</div> `,
})
class MockFaIconComponent {
  @Input() icon: string | string[] | undefined;
}

describe('ControlsComponent', () => {
  let component: ControlsComponent;
  let fixture: ComponentFixture<ControlsComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ControlsComponent],
      providers: [
        {
          provide: ScreenService,
          useValue: { observerBreakpoints: () => ({ pipe: jest.fn() }) },
        },
        { provide: FaIconLibrary, useValue: { addIcons: jest.fn() } },
      ],
    })
      .overrideComponent(ControlsComponent, {
        set: { imports: [NgIf, AsyncPipe, MatButtonModule, MockFaIconComponent] },
      })
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ControlsComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      // expect(fixture).toMatchSnapshot();
    });
  });

  describe('ngOnDestroy', () => {
    it('should call destroy method', () => {
      // prepare
      jest.spyOn(component.clipboard, 'destroy');

      // act
      component.ngOnDestroy();

      // check
      expect(component.clipboard.destroy).toHaveBeenCalled();
    });
  });

  describe('refreshClickHandler', () => {
    it('should emit refreshClick', () => {
      // prepare
      jest.spyOn(component.refreshClick, 'emit');

      // act
      component.refreshClickHandler();

      // check
      expect(component.refreshClick.emit).toHaveBeenCalled();
    });
  });

  describe('copyNumberClickHandler', () => {
    it('should emit refreshClick', () => {
      // prepare
      const message: string = 'Nummer gekopieerd naar klembord';
      jest.spyOn(component.copyNumberClick, 'emit');

      // act
      component.copyNumberClickHandler();

      // check
      expect(component.copyNumberClick.emit).toHaveBeenCalledWith(message);
    });
  });

  describe('copyOgmClickHandler', () => {
    it('should emit refreshClick', () => {
      // prepare
      const message: string = 'Mededeling gekopieerd naar klembord';
      jest.spyOn(component.copyOgmClick, 'emit');

      // act
      component.copyOgmClickHandler();

      // check
      expect(component.copyOgmClick.emit).toHaveBeenCalledWith(message);
    });
  });
});
