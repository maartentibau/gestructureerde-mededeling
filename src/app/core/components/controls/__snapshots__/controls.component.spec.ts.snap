// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsComponent rendering component should match snapshot 1`] = `
<ogm-controls
  clipboard={[Function Clipboard]}
  copyNumber={[Function Function]}
  copyNumberClick={[Function OutputEmitterRef]}
  copyOgm={[Function Function]}
  copyOgmClick={[Function OutputEmitterRef]}
  isSmallScreen$="undefined"
  ogm={[Function Function]}
  refresh={[Function Function]}
  refreshClick={[Function OutputEmitterRef]}
>
  <div
    class="controls large"
  >
  </div>
</ogm-controls>
`;
